src/services/analytics/TagManager.ts

```diff
@@ -1,94 +1,114 @@
-      window.${DATA_LAYER_NAME}.push(${JSON.stringify(dataLayer)})`
+// https://developers.google.com/tag-platform/devguides/privacy#turn_off_google_analytics
+const GA_DISABLE_KEY = `ga-disable-${GOOGLE_ANALYTICS_MEASUREMENT_ID}`
+const GTM_DISABLE_TRIGGER_COOKIE_NAME = 'google-analytics-opt-out'
 
-  script.innerHTML = gtmDataLayerScript
-
-  return script
-}
+// Injected GTM script singleton
+let gtmScriptRef: HTMLScriptElement | null = null
 
 const TagManager = {
+  isEnabled: () => {
+    // @ts-expect-error - Element implicitly has an 'any' type because index expression is not of type 'number'.
+    return Cookies.get(GTM_DISABLE_TRIGGER_COOKIE_NAME) === undefined && window[GA_DISABLE_KEY] === false
+  },
   initialize: (args: TagManagerArgs) => {
-    const { dataLayer } = _getRequiredGtmArgs(args)
+    if (TagManager.isEnabled()) {
+      return
+    }
+
+    // Enable GA
+    // @ts-expect-error - Element implicitly has an 'any' type because index expression is not of type 'number'.
+    window[GA_DISABLE_KEY] = false
 
-    if (dataLayer) {
-      const gtmDataLayerScript = _getGtmDataLayerScript(dataLayer)
-      document.head.appendChild(gtmDataLayerScript)
+    // Enable GTM triggers
+    Cookies.remove(GTM_DISABLE_TRIGGER_COOKIE_NAME, { path: '/' })
+
+    if (gtmScriptRef) {
+      return
     }
 
-    const gtmScript = _getGtmScript(args)
-    document.head.insertBefore(gtmScript, document.head.childNodes[0])
+    // Initialize dataLayer (with configuration)
+    window[DATA_LAYER_NAME] = args.dataLayer ? [args.dataLayer] : []
+
+    gtmScriptRef = _getGtmScript(args)
+
+    // Initialize GTM. This pushes the default dataLayer event:
+    // { "gtm.start": new Date().getTime(), event: "gtm.js" }
+    document.head.insertBefore(gtmScriptRef, document.head.childNodes[0])
   },
   dataLayer: (dataLayer: DataLayer) => {
-    if (window[DATA_LAYER_NAME]) {
-      return window[DATA_LAYER_NAME].push(dataLayer)
+    if (!TagManager.isEnabled()) {
+      return
+    }
+
+    window[DATA_LAYER_NAME].push(dataLayer)
+
+    if (!IS_PRODUCTION) {
+      console.info('[GTM] -', dataLayer)
     }
+  },
+  disable: () => {
+    if (!TagManager.isEnabled()) {
+      return
+    }
+
+    // Disable GA
+    // @ts-expect-error - Element implicitly has an 'any' type because index expression is not of type 'number'.
+    window[GA_DISABLE_KEY] = true
+
+    // Disable GTM triggers
+    Cookies.set(GTM_DISABLE_TRIGGER_COOKIE_NAME, 'true', {
+      expires: Number.MAX_SAFE_INTEGER,
+      path: '/',
+    })
+
+    // const GTM_COOKIE_LIST = ['_ga', '_gat', '_gid']
 
-    const gtmDataLayerScript = _getGtmDataLayerScript(dataLayer)
-    document.head.insertBefore(gtmDataLayerScript, document.head.childNodes[0])
+    // GTM_COOKIE_LIST.forEach((cookie) => {
+    //   Cookies.remove(cookie, {
+    //     path: '/',
+    //     domain: `.${location.host.split('.').slice(-2).join('.')}`,
+    //   })
+    // })
```

Comment:
These need not be removed anymore but we need to determine if it's legal to keep them.